File: lib/alttext_api.rb
require "net/http"
require "uri"
require "json"

class AltTextApiError < StandardError
  attr_reader :status, :error_code, :errors, :raw_body

  def initialize(status:, error_code:, errors:, raw_body:, message:)
    @status = status
    @error_code = error_code
    @errors = errors
    @raw_body = raw_body
    super(message)
  end
end

class AltTextApi
  def initialize(api_key:, base_url: nil)
    @api_key = api_key
    @base_url = (base_url || "https://alttext.ai/api/v1").chomp("/")
  end

  def get_account
    data, = request(:get, "/account")
    data
  end

  def list_images(page: nil, limit: nil, lang: nil)
    query = {}
    query[:page] = page if page
    query[:limit] = limit if limit
    query[:lang] = lang if lang

    data, headers = request(:get, "/images", query: query)
    {
      images: data["images"],
      pagination: parse_pagination(headers),
    }
  end

  def search_images(query:, limit: nil, lang: nil)
    params = { q: query }
    params[:limit] = limit if limit
    params[:lang] = lang if lang

    data, headers = request(:get, "/images/search", query: params)
    {
      images: data["images"],
      pagination: parse_pagination(headers),
    }
  end

  def get_image(asset_id:, lang: nil)
    query = {}
    query[:lang] = lang if lang

    data, = request(:get, "/images/#{URI.encode_uri_component(asset_id)}", query: query)
    data
  end

  def create_image(url:, asset_id: nil, lang: nil, keywords: nil, negative_keywords: nil,
                   gpt_prompt: nil, max_chars: nil, overwrite: nil, tags: nil, metadata: nil)
    image_envelope = { url: url }
    image_envelope[:asset_id] = asset_id if asset_id
    image_envelope[:tags] = tags if tags
    image_envelope[:metadata] = metadata if metadata

    body = { image: image_envelope, async: false }
    body[:lang] = lang if lang
    body[:keywords] = keywords if keywords
    body[:negative_keywords] = negative_keywords if negative_keywords
    body[:gpt_prompt] = gpt_prompt if gpt_prompt
    body[:max_chars] = max_chars if max_chars
    body[:overwrite] = overwrite unless overwrite.nil?

    data, = request(:post, "/images", body: body)
    data
  end

  def update_image(asset_id:, alt_text: nil, tags: nil, metadata: nil, lang: nil, overwrite: nil)
    image_envelope = {}
    image_envelope[:alt_text] = alt_text unless alt_text.nil?
    image_envelope[:tags] = tags if tags
    image_envelope[:metadata] = metadata if metadata

    body = { image: image_envelope }
    body[:lang] = lang if lang
    body[:overwrite] = overwrite unless overwrite.nil?

    data, = request(:patch, "/images/#{URI.encode_uri_component(asset_id)}", body: body)
    data
  end

  def delete_image(asset_id:)
    data, = request(:delete, "/images/#{URI.encode_uri_component(asset_id)}")
    data
  end

  def scrape_page(url:, html: nil, include_existing: nil, lang: nil, keywords: nil,
                  negative_keywords: nil, gpt_prompt: nil, max_chars: nil, overwrite: nil)
    page_scrape_envelope = { url: url }
    page_scrape_envelope[:html] = html if html

    body = { page_scrape: page_scrape_envelope }
    body[:include_existing] = include_existing unless include_existing.nil?
    body[:lang] = lang if lang
    body[:keywords] = keywords if keywords
    body[:negative_keywords] = negative_keywords if negative_keywords
    body[:gpt_prompt] = gpt_prompt if gpt_prompt
    body[:max_chars] = max_chars if max_chars
    body[:overwrite] = overwrite unless overwrite.nil?

    data, = request(:post, "/images/page_scrape", body: body)
    data
  end

  private

  def request(method, path, body: nil, query: nil)
    uri = URI("#{@base_url}#{path}")
    if query&.any?
      uri.query = URI.encode_www_form(query.compact)
    end

    http = Net::HTTP.new(uri.host, uri.port)
    http.use_ssl = uri.scheme == "https"
    http.open_timeout = 10
    http.read_timeout = 120

    req = build_request(method, uri, body)
    response = http.request(req)

    handle_response(response)
  rescue Errno::ECONNREFUSED, Net::OpenTimeout, Net::ReadTimeout, SocketError => e
    raise AltTextApiError.new(
      status: 0,
      error_code: "connection_error",
      errors: {},
      raw_body: {},
      message: "Could not connect to AltText.ai API: #{e.message}",
    )
  end

  def build_request(method, uri, body)
    klass = case method
            when :get    then Net::HTTP::Get
            when :post   then Net::HTTP::Post
            when :patch  then Net::HTTP::Patch
            when :delete then Net::HTTP::Delete
            end

    req = klass.new(uri)
    req["X-API-Key"] = @api_key
    req["Accept"] = "application/json"

    if body
      req["Content-Type"] = "application/json"
      req.body = JSON.generate(body)
    end

    req
  end

  def handle_response(response)
    body = begin
      JSON.parse(response.body || "{}")
    rescue JSON::ParserError
      {}
    end

    unless response.is_a?(Net::HTTPSuccess)
      error_code = body["error_code"]
      errors = body["errors"] || {}
      message = body["error"] ||
                errors.values.flatten.join(", ").then { |s| s.empty? ? nil : s } ||
                "HTTP #{response.code}"

      raise AltTextApiError.new(
        status: response.code.to_i,
        error_code: error_code,
        errors: errors,
        raw_body: body,
        message: message,
      )
    end

    [body, response]
  end

  def parse_pagination(response)
    {
      current_page: response["current-page"]&.to_i || 1,
      page_items: response["page-items"]&.to_i || 20,
      total_pages: response["total-pages"]&.to_i || 1,
      total_count: response["total-count"]&.to_i || 0,
    }
  end
end

File: src/alttext-api.ts
import type {
  AccountRecord,
  BulkCreateResult,
  GenerateOptions,
  ImageListResult,
  ImageRecord,
  Pagination,
  ScrapePageOptions,
  ScrapeResult,
  UpdateAccountOptions,
  UpdateImageOptions,
} from "./types.js";

const REQUEST_TIMEOUT_MS = 120_000;

type HttpMethod = "GET" | "POST" | "PATCH" | "DELETE";

/** Structured error from the AltText.ai API. */
export class AltTextApiError extends Error {
  readonly status: number;
  readonly errorCode: string | null;
  readonly errors: Readonly<Record<string, readonly string[]>>;

  constructor(opts: {
    status: number;
    errorCode: string | null;
    errors: Record<string, string[]>;
    message: string;
  }) {
    super(opts.message);
    this.name = "AltTextApiError";
    this.status = opts.status;
    this.errorCode = opts.errorCode;
    this.errors = opts.errors;
  }
}

/** HTTP client for the AltText.ai REST API. */
export class AltTextApi {
  private readonly apiKey: string;
  private readonly baseUrl: string;

  constructor(apiKey: string, baseUrl?: string) {
    this.apiKey = apiKey;
    this.baseUrl = (baseUrl ?? "https://alttext.ai/api/v1").replace(/\/+$/, "");
  }

  async getAccount(): Promise<AccountRecord> {
    const { data } = await this.request("GET", "/account");
    return data as unknown as AccountRecord;
  }

  async updateAccount(opts: UpdateAccountOptions): Promise<AccountRecord> {
    const envelope: Record<string, unknown> = {};
    if (opts.name !== undefined) envelope["name"] = opts.name;
    if (opts.webhookUrl !== undefined) envelope["webhook_url"] = opts.webhookUrl;
    if (opts.notificationEmail !== undefined)
      envelope["notification_email"] = opts.notificationEmail;

    const { data } = await this.request("PATCH", "/account", {
      body: { account: envelope },
    });
    return data as unknown as AccountRecord;
  }

  async listImages(
    opts: { page?: number; limit?: number; lang?: string } = {},
  ): Promise<ImageListResult> {
    const query = this.buildQuery({
      page: opts.page,
      limit: opts.limit,
      lang: opts.lang,
    });

    const { data, response } = await this.request("GET", "/images", { query });
    const typed = data as unknown as { images: ImageRecord[] };
    return {
      images: typed.images,
      pagination: parsePagination(response),
    };
  }

  async searchImages(opts: {
    query: string;
    limit?: number;
    lang?: string;
  }): Promise<ImageListResult> {
    const query = this.buildQuery({
      q: opts.query,
      limit: opts.limit,
      lang: opts.lang,
    });

    const { data, response } = await this.request("GET", "/images/search", { query });
    const typed = data as unknown as { images: ImageRecord[] };
    return {
      images: typed.images,
      pagination: parsePagination(response),
    };
  }

  async getImage(opts: { assetId: string; lang?: string }): Promise<ImageRecord> {
    const query = this.buildQuery({ lang: opts.lang });
    const { data } = await this.request("GET", `/images/${encodeURIComponent(opts.assetId)}`, {
      query,
    });
    return data as unknown as ImageRecord;
  }

  async createImage(opts: { url: string } & GenerateOptions): Promise<ImageRecord> {
    return this.generateImage({ url: opts.url }, opts);
  }

  async createImageFromRaw(opts: { raw: string } & GenerateOptions): Promise<ImageRecord> {
    return this.generateImage({ raw: opts.raw }, opts);
  }

  async translateImage(opts: { assetId: string; lang: string }): Promise<ImageRecord> {
    const body = {
      image: { asset_id: opts.assetId },
      lang: opts.lang,
      async: false,
    };
    const { data } = await this.request("POST", "/images", { body });
    return data as unknown as ImageRecord;
  }

  async updateImage(opts: UpdateImageOptions): Promise<ImageRecord> {
    const imageEnvelope: Record<string, unknown> = {};
    if (opts.altText !== undefined) imageEnvelope["alt_text"] = opts.altText;
    if (opts.tags !== undefined) imageEnvelope["tags"] = opts.tags;
    if (opts.metadata !== undefined) imageEnvelope["metadata"] = opts.metadata;

    const body: Record<string, unknown> = { image: imageEnvelope };
    if (opts.lang !== undefined) body["lang"] = opts.lang;
    if (opts.overwrite !== undefined) body["overwrite"] = opts.overwrite;

    const { data } = await this.request("PATCH", `/images/${encodeURIComponent(opts.assetId)}`, {
      body,
    });
    return data as unknown as ImageRecord;
  }

  async deleteImage(opts: { assetId: string }): Promise<void> {
    await this.request("DELETE", `/images/${encodeURIComponent(opts.assetId)}`);
  }

  async bulkCreate(opts: { csvFile: Blob; email?: string }): Promise<BulkCreateResult> {
    const formData = new FormData();
    formData.append("file", opts.csvFile);
    if (opts.email) formData.append("email", opts.email);

    const url = `${this.baseUrl}/images/bulk_create`;
    let response: Response;
    try {
      response = await fetch(url, {
        method: "POST",
        headers: { "X-API-Key": this.apiKey, Accept: "application/json" },
        body: formData,
        signal: AbortSignal.timeout(REQUEST_TIMEOUT_MS),
      });
    } catch (err) {
      throw connectionError(err);
    }

    return handleResponse(response) as Promise<BulkCreateResult>;
  }

  async scrapePage(opts: ScrapePageOptions): Promise<ScrapeResult> {
    const pageScrapeEnvelope: Record<string, unknown> = { url: opts.url };
    if (opts.html !== undefined) pageScrapeEnvelope["html"] = opts.html;

    const body: Record<string, unknown> = { page_scrape: pageScrapeEnvelope };
    if (opts.includeExisting !== undefined) body["include_existing"] = opts.includeExisting;
    if (opts.lang !== undefined) body["lang"] = opts.lang;
    if (opts.keywords !== undefined) body["keywords"] = opts.keywords;
    if (opts.negativeKeywords !== undefined) body["negative_keywords"] = opts.negativeKeywords;
    if (opts.gptPrompt !== undefined) body["gpt_prompt"] = opts.gptPrompt;
    if (opts.maxChars !== undefined) body["max_chars"] = opts.maxChars;
    if (opts.overwrite !== undefined) body["overwrite"] = opts.overwrite;

    const { data } = await this.request("POST", "/images/page_scrape", { body });
    return data as unknown as ScrapeResult;
  }

  // -- Private helpers --

  private async generateImage(
    imageSource: Record<string, unknown>,
    opts: GenerateOptions,
  ): Promise<ImageRecord> {
    const imageEnvelope = { ...imageSource };
    if (opts.assetId !== undefined) imageEnvelope["asset_id"] = opts.assetId;
    if (opts.tags !== undefined) imageEnvelope["tags"] = opts.tags;
    if (opts.metadata !== undefined) imageEnvelope["metadata"] = opts.metadata;

    const body: Record<string, unknown> = { image: imageEnvelope, async: false };
    if (opts.lang !== undefined) body["lang"] = opts.lang;
    if (opts.keywords !== undefined) body["keywords"] = opts.keywords;
    if (opts.negativeKeywords !== undefined) body["negative_keywords"] = opts.negativeKeywords;
    if (opts.gptPrompt !== undefined) body["gpt_prompt"] = opts.gptPrompt;
    if (opts.maxChars !== undefined) body["max_chars"] = opts.maxChars;
    if (opts.overwrite !== undefined) body["overwrite"] = opts.overwrite;

    const { data } = await this.request("POST", "/images", { body });
    return data as unknown as ImageRecord;
  }

  private async request(
    method: HttpMethod,
    path: string,
    opts: { body?: Record<string, unknown>; query?: Record<string, string> } = {},
  ): Promise<{ data: Record<string, unknown>; response: Response }> {
    let url = `${this.baseUrl}${path}`;
    if (opts.query && Object.keys(opts.query).length > 0) {
      url += `?${new URLSearchParams(opts.query).toString()}`;
    }

    const headers: Record<string, string> = {
      "X-API-Key": this.apiKey,
      Accept: "application/json",
    };

    const fetchOpts: RequestInit = {
      method,
      headers,
      signal: AbortSignal.timeout(REQUEST_TIMEOUT_MS),
    };

    if (opts.body) {
      headers["Content-Type"] = "application/json";
      fetchOpts.body = JSON.stringify(opts.body);
    }

    let response: Response;
    try {
      response = await fetch(url, fetchOpts);
    } catch (err) {
      throw connectionError(err);
    }

    const data = await handleResponse(response);
    return { data, response };
  }

  private buildQuery(params: Record<string, string | number | undefined>): Record<string, string> {
    const query: Record<string, string> = {};
    for (const [key, value] of Object.entries(params)) {
      if (value !== undefined) query[key] = String(value);
    }
    return query;
  }
}

// -- Module-level helpers --

async function handleResponse(response: Response): Promise<Record<string, unknown>> {
  let body: Record<string, unknown>;
  try {
    body = (await response.json()) as Record<string, unknown>;
  } catch {
    body = {};
  }

  if (!response.ok) {
    const errorCode = (body["error_code"] as string | undefined) ?? null;
    const rawErrors = body["errors"];
    const errors: Record<string, string[]> = {};
    if (typeof rawErrors === "object" && rawErrors !== null && !Array.isArray(rawErrors)) {
      for (const [key, value] of Object.entries(rawErrors as Record<string, unknown>)) {
        if (Array.isArray(value) && value.every((v): v is string => typeof v === "string")) {
          errors[key] = value;
        }
      }
    }

    const errorValues = Object.values(errors).flat();
    const message =
      (body["error"] as string | undefined) ??
      (errorValues.length > 0 ? errorValues.join(", ") : null) ??
      `HTTP ${String(response.status)}`;

    throw new AltTextApiError({ status: response.status, errorCode, errors, message });
  }

  return body;
}

function safeParseInt(value: string | null, fallback: number): number {
  const parsed = parseInt(value ?? "", 10);
  return Number.isNaN(parsed) ? fallback : parsed;
}

function parsePagination(response: Response): Pagination {
  return {
    currentPage: safeParseInt(response.headers.get("current-page"), 1),
    pageItems: safeParseInt(response.headers.get("page-items"), 20),
    totalPages: safeParseInt(response.headers.get("total-pages"), 1),
    totalCount: safeParseInt(response.headers.get("total-count"), 0),
  };
}

function connectionError(err: unknown): AltTextApiError {
  const message = err instanceof Error ? err.message : String(err);
  return new AltTextApiError({
    status: 0,
    errorCode: "connection_error",
    errors: {},
    message: `Could not connect to AltText.ai API: ${message}`,
  });
}
